pkg/ddl/backfilling.go

```diff
@@ -427,79 +426,81 @@ func (w *backfillWorker) run(d *ddlCtx, bf backfiller, job *model.Job) {
 	}
 }
 
-func splitAndValidateTableRanges(
+// loadTableRanges load table key ranges from PD between given start key and end key.
+// It returns up to `limit` ranges.
+func loadTableRanges(
 	ctx context.Context,
 	t table.PhysicalTable,
 	store kv.Storage,
 	startKey, endKey kv.Key,
 	limit int,
 ) ([]kv.KeyRange, error) {
-	ranges, err := splitTableRanges(ctx, t, store, startKey, endKey, limit)
-	if err != nil {
-		return nil, err
-	}
-	return validateTableRanges(ranges, startKey, endKey)
-}
-
-func validateTableRanges(ranges []kv.KeyRange, start, end kv.Key) ([]kv.KeyRange, error) {
-	for i, r := range ranges {
-		if len(r.StartKey) == 0 {
-			if i != 0 {
-				return nil, errors.Errorf("get empty start key in the middle of ranges")
-			}
-			r.StartKey = start
-		}
-		if len(r.EndKey) == 0 {
-			if i != len(ranges)-1 {
-				return nil, errors.Errorf("get empty end key in the middle of ranges")
-			}
-			r.EndKey = end
-		}
-	}
-	return ranges, nil
-}
-
-// splitTableRanges uses PD region's key ranges to split the backfilling table key range space,
-// to speed up backfilling data in table with disperse handle.
-// The `t` should be a non-partitioned table or a partition.
-func splitTableRanges(
-	ctx context.Context,
-	t table.PhysicalTable,
-	store kv.Storage,
-	startKey, endKey kv.Key,
-	limit int,
-) ([]kv.KeyRange, error) {
-	logutil.DDLLogger().Info("split table range from PD",
+	logutil.DDLLogger().Info("load table ranges from PD",
 		zap.Int64("physicalTableID", t.GetPhysicalID()),
 		zap.String("start key", hex.EncodeToString(startKey)),
 		zap.String("end key", hex.EncodeToString(endKey)))
 	if len(startKey) == 0 && len(endKey) == 0 {
-		logutil.DDLLogger().Info("split table range from PD, get noop table range",
+		logutil.DDLLogger().Info("load table range from PD, get noop table range",
 			zap.Int64("physicalTableID", t.GetPhysicalID()))
 		return []kv.KeyRange{}, nil
 	}
 
-	kvRange := kv.KeyRange{StartKey: startKey, EndKey: endKey}
 	s, ok := store.(tikv.Storage)
 	if !ok {
 		// Only support split ranges in tikv.Storage now.
-		return []kv.KeyRange{kvRange}, nil
+		return []kv.KeyRange{{StartKey: startKey, EndKey: endKey}}, nil
 	}
 
+	rc := s.GetRegionCache()
 	maxSleep := 10000 // ms
-	bo := backoff.NewBackofferWithVars(ctx, maxSleep, nil)
-	rc := copr.NewRegionCache(s.GetRegionCache())
-	ranges, err := rc.SplitRegionRanges(bo, []kv.KeyRange{kvRange}, limit)
+	bo := tikv.NewBackofferWithVars(ctx, maxSleep, nil)
+	rs, err := rc.BatchLoadRegionsWithKeyRange(bo, startKey, endKey, limit)
```

Comment:
I check its logic with my lightning knowledge. In
BatchLoadRegionsWithKeyRange -> scanRegions -> handleRegionInfos (needLeader is `true`), seems if the region from PD does not contains leader store, the region will not return to caller. You have added a `validateAndFillRanges` function to find this case.

This defect of scan region API causes many duplicated logic, like lightning/BR's `checkRegionConsistency`. I have asked PD dev to move these checking logic into PD client.

lgtm for this PR, let's push PD dev @okJiang